chat-history: Add MessageAnimation widget

melix99 opened the pull request:
Requires #165 merged first. I wanted to see if animations (AKA gifs or muted videos) work correctly with the new 'Media' widget. They do apparently, with just a small modification in the MediaPicture.

melix99 commented on src/session/content/message_row/media_paintable.rs:
```diff
@@ -0,0 +1,128 @@
+
+        let self_ = imp::MediaPaintable::from_instance(self);
+        if let Some(ref paintable) = paintable {
+            paintable.connect_invalidate_contents(clone!(@weak self as obj => move |_| {
```
This signal should be disconnected when the paintable is changed.

marhkb commented on src/session/content/message_row/animation.rs:
```diff
@@ -0,0 +1,158 @@
+impl MessageAnimation {
+    pub fn new(message: &glib::Object) -> Self {
+        glib::Object::new(&[("message", message)]).expect("Failed to create MessageAnimation")
+    }
```
```suggestion
```

marhkb commented on README.md:
```diff
@@ -49,6 +49,8 @@ The following packages are required to build Telegrand:
 - TDLib 1.8.0
 - [Telegram API Credentials](https://my.telegram.org/) (optional, but recommended)
 
+Also, Telegrand requires the Libav GStreamer plugin installed in your system to correctly show media files.
```
We really need other people to check whether that is enough. Here on Fedora 35, at least, it seems I need `gstreamer1-plugins-base` `gstreamer1-plugins-good`, `gstreamer1-plugins-bad-free` in addition to `gstreamer1-libav` from rpmfusion. If one of them is missing the animation is not shown.

melix99 replied:
Mh, weird, at least by looking at gst docs it seems that GIF and MPEG4 decoding are handled by the libav plugin. I'll investigate a bit more.

melix99 replied:
I finally tested on arch and it seems that they are all required indeed.

melix99 replied:
I checked again and it seems that now I only need `gstreamer-plugins-good` and `gstreamer-libav` installed.

marhkb commented on src/session/content/message_row/media_picture.rs:
```diff
@@ -162,8 +164,22 @@ impl MediaPicture {
 
         let self_ = imp::MediaPicture::from_instance(self);
+        if let Some(old_paintable) = self_.paintable.borrow().as_ref() {
+            if let Some(id) = self_.handler_id.take() {
```
In this case the inner `if let` statement is obsolete. The `handler_id` must never be `None` if the paintable is `Some`.
It can be directly unwrapped.

melix99 commented on src/session/content/message_row/media.rs:
```diff
@@ -34,159 +37,136 @@ mod imp {
-        if self.caption() == caption {
-            return;
-        }
+    fn download_media(&self, file_id: i32, session: &Session) {
```
We should probably have a way to stop the download when the message of this widget changes, otherwise we risk to end up with the wrong media to be loaded.

marhkb replied:
We could save the `file_id` before we start the download and in `receiver.attach` we compare it with that `file_id` we get in the closure. Then, we return `glib::Continue(false)` if the `file_id` isn't the same anymore.

melix99 replied:
This actually needs further changes with file downloading in Telegrand, so I'll postpone this issue for now.

marhkb commented on src/session/content/message_row/media.rs:
```diff
@@ -34,159 +37,136 @@ mod imp {
+        if let Some(old_message) = imp.old_message.upgrade() {
+            if let Some(id) = imp.handler_id.take() {
+                old_message.disconnect(id);
             }
```
Again, I think explicitly testing for `Some` isn't necessary here, as the handler should always be `Some`.
Also, couldn't the unwatching of the binding not put in `old_message` `if let` block, too? It could also be unwrapped directly.
This would remove the cyclomatic complexity a bit.

marhkb commented on src/session/content/message_row/media.rs:
```diff
@@ -34,159 +37,136 @@ mod imp {
+                        }
+                        MessageContent::MessagePhoto(data) => parse_formatted_text(data.caption),
+                        _ => unreachable!(),
+                    }
```
```suggestion
                    parse_formatted_text(match content.0 {
                        MessageContent::MessageAnimation(data) => data.caption,
                        MessageContent::MessagePhoto(data) => data.caption,
                        _ => unreachable!(),
                    })
```

marhkb commented on src/session/content/message_row/media.rs:
```diff
@@ -34,159 +37,136 @@ mod imp {
+                MessageContent::MessageAnimation(_) => {
+                    let media_file = gtk::MediaFile::for_filename(&path);
+                    media_file.set_loop(true);
+                    media_file.play();
```
Not relevant for this PR: In the future, we should consider pause this when the application window has no focus, for example.

melix99 replied:
Yeah, maybe. Even though it would be better pausing the non-visible animations first, but currently I don't know how to achieve that.

marhkb commented on src/session/content/message_row/media.rs:
```diff
@@ -34,159 +37,136 @@ mod imp {
+                        .set_paintable(Some(gdk::Texture::from_file(&file).unwrap().upcast()));
+                }
+                _ => unreachable!(),
+            }
```
```suggestion
            self.imp().content.set_paintable(match message.content().0 {
                MessageContent::MessageAnimation(_) => {
                    let media_file = gtk::MediaFile::for_filename(&path);
                    media_file.set_loop(true);
                    media_file.play();
                    Some(media_file.upcast())
                }
                MessageContent::MessagePhoto(_) => {
                    // TODO: Consider changing this to use an async api when
                    // https://github.com/gtk-rs/gtk4-rs/pull/777 is merged
                    let file = gio::File::for_path(path);
                    Some(gdk::Texture::from_file(&file).unwrap().upcast())
                }
                _ => unreachable!(),
            });
```

marhkb commented on data/resources/ui/content-message-media-content.ui:
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<interface>
+  <template class="ContentMessageMediaContent" parent="GtkWidget">
```
The name is a bit confusing, with having "Content" in the beginning and in the end. Have you considered calling it `ContentMessageMediaEnvelope` or `ContentMessageMediaFrame`?

melix99 replied:
I like the "Content" at the end because the widget is literally the MessageMedia's content property. But I agree that the prefix makes it weird, but honestly I think the problem is the naming scheme we're using that is a bit weird. I was thinking of dropping all the directory-based prefixes and just using "Telegrand", like any other gtk app. E.g. this would be "TelegrandMessageMediaContent" with that change. But that's something to be done separately.

marhkb commented on data/resources/ui/content-message-media.ui:
```diff
@@ -1,49 +1,8 @@
+  <template class="ContentMessageMedia" parent="ContentMessageRow">
+    <property name="content">
+      <object class="ContentMessageMediaContent" id="content"/>
+    </property>
```
What was your intention in not unifying this file with `content-media-message-content.ui`?

melix99 replied:
The `ContentMessageMediaContent` needs a custom sizing implementation, so I had to create a separate custom widget just for that reason.

marhkb replied:
Oh yes, I think you told that once. I hope I don't ask that again in a week. :-p

melix99 committed:
media-picture: Merge it with MessageMedia

Also stop showing the progress bar, since it's usually shown only for a
split second. Instead, with this commit the minithumbnail is shown while
the photo is downloading.

melix99 committed:
chat-history: Add MessageAnimation widget
